src/v/cluster/node_status_provider.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/members_table.h"
+#include "cluster/node_status_provider_rpc_service.h"
+#include "config/property.h"
+#include "rpc/connection_cache.h"
+#include "seastarx.h"
+#include "ssx/metrics.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/timer.hh>
+
+#include <absl/container/flat_hash_set.h>
+
+namespace cluster {
+
+struct node_status {
+    model::node_id node_id;
+    rpc::clock_type::time_point last_seen;
+};
+
+/*
+ * node_status_provider is a sub-system that operates outside of Raft
+ * and has the purpose of tracking the status of all peers. Its operation is as
+ * follows:
+ * 1. Maintain a list of peers for this node. This is currently done via a
+ * callback from the members_table.
+ * 2. Send a periodic node_status RPC to all known peers and record the results.
+ *
+ * It runs only on shard 0.
+ */
+class node_status_provider {
+public:
+    static constexpr ss::shard_id shard = 0;
+
+    // TODO: Add seed servers and Raft snapshot here
+    node_status_provider(
+      model::node_id,
+      members_table&,
+      config::binding<std::chrono::milliseconds>);
+
+    void start();
+    ss::future<> stop();
+
+    std::optional<node_status> get_node_status(model::node_id) const;
+
+private:
+    void handle_members_updated_notification(std::vector<model::node_id>);
+
+    void update_node_status(node_status_metadata&&);
```

Comment:
do we need an rvalue refrence here ?

Reply:
It forces the caller to move the status. Currently, we don't gain anything from it as the status is a trivial type, but that change in the future. I can change it if you feel there's no point to it.

Reply:
Changed to take by value.